fedimint-atomic-broadcast/src/session.rs

```diff
@@ -0,0 +1,242 @@
+use std::collections::BTreeMap;
+use std::time::Duration;
+
+use aleph_bft::Keychain as KeychainTrait;
+use async_channel::{Receiver, Sender};
+use bitcoin_hashes::{sha256, Hash};
+use fedimint_core::encoding::Encodable;
+use fedimint_core::task::sleep;
+use tokio::sync::{oneshot, watch};
+
+use crate::conversion::{to_node_index, to_peer_id};
+use crate::data_provider::{DataProvider, UnitData};
+use crate::finalization_handler::FinalizationHandler;
+use crate::keychain::Keychain;
+use crate::network::Network;
+use crate::spawner::Spawner;
+use crate::{db, AcceptedItem, Block, Decision, Message, OrderedItem, Recipient, SignedBlock};
+
+// this function completes a session with the following steps:
+// - set up the aleph bft session
+// - periodically request the corresponding signed block from peers
+// - combine the accepted items into a block until we reach a preset number of
+//   ordered batches or a signed block arrives
+// - collect signatures until we reach the threshold or a signed block arrives
+#[allow(clippy::too_many_arguments)]
+pub async fn run(
+    index: u64,
+    keychain: Keychain,
+    backup_loader: std::io::Cursor<Vec<u8>>,
+    backup_saver: db::UnitSaver,
+    item_receiver: Receiver<Vec<u8>>,
+    network_data_receiver: Receiver<Vec<u8>>,
+    outgoing_message_sender: Sender<(Message, Recipient)>,
+    ordered_item_sender: tokio::sync::mpsc::Sender<OrderedItem>,
+    signed_block_receiver: Receiver<SignedBlock>,
+) -> anyhow::Result<SignedBlock> {
+    const MAX_ROUND: u16 = 5000;
+    const ROUND_DELAY: f64 = 250.0;
+    const EXPONETIAL_SLOWDOWN_OFFSET: usize = 3000;
+    const BASE: f64 = 1.01;
+    const BLOCK_REQUEST_DELAY: Duration = Duration::from_secs(10);
+
+    let mut config = aleph_bft::default_config(
+        keychain.peer_count.into(),
+        keychain.peer_id.into(),
+        index,
+    );
+
+    // In order to bound a sessions RAM consumption we need to bound its number of
+    // units and therefore its number of rounds. Since we use a session to
+    // create a threshold signature for the corresponding block we have to
+    // guarantee that an attacker cannot exhaust our memory by preventing the
+    // creation of a threshold signature, thereby keeping the session open
+    // indefinitely. Hence we increase the delay between rounds exponentially
+    // such that MAX_ROUND would only be reached after roughly 350 years.
+    // In case of such an attack the broadcast stops ordering any items until the
+    // attack subsides.
+    config.max_round = MAX_ROUND;
+    config.delay_config.unit_creation_delay = std::sync::Arc::new(|round_index| {
+        let delay = if round_index == 0 {
+            0.0
+        } else if round_index < EXPONETIAL_SLOWDOWN_OFFSET {
```

Comment:
Wouldn't this branch disappear if `round_index.saturating_sub(EXPONETIAL_SLOWDOWN_OFFSET)` or something like that be used below?